pkg/auditlog/auditlog.go

```diff
@@ -0,0 +1,88 @@
+// Package alog is a package to send audit logs events to WorkOS.
+//
+// Example:
+//   func main() {
+//       alog.SetAPIKey("my_api_key")
+//       defer alog.Close()
+//
+//       // Wherever you need to publish an audit log event:
+//       alog.Publish(alog.Event{
+//           Action:     "document.viewed",
+//           ActionType: "r",
+//           ActorName:  "Jairo Kunde",
+//           ActorID:    "user_01DGZ0FAXN978HCET66Q98QMTQ",
+//           Group:      "abstract.com",
+//           Location:   "55.27.223.26",
+//           OccurredAt: time.Now(),
+//           TargetName: "central.class",
+//           TargetID:   "doc_01DGZ0FAXP4HA4X0BVFKS0ZH4Y",
+//       })
+//   }
+package alog
+
+import (
+	"os"
+	"time"
+)
+
+var (
+	// DefaultPublisher is the publisher used by SetAPIKey, Publish and Close
+	// functions.
+	DefaultPublisher = &Publisher{
+		APIKey:    os.Getenv("WORKOS_API_KEY"),
+		Endpoint:  "https://api.workos.com/events",
+		QueueSize: 512,
+	}
+
+	currentLocation string
+)
+
+func init() {
+	currentLocation, _ = os.Hostname()
+}
+
+// SetAPIKey sets the WorkOS API key to use when using Publish.
+func SetAPIKey(k string) {
+	DefaultPublisher.APIKey = k
+}
+
+// Publish publishes the given events.
+func Publish(events ...Event) {
+	DefaultPublisher.Publish(events...)
+}
+
+// Close stops publishings audit log events and releases allocated resources.
+// It waits for pending events to be sent before returning.
+func Close() {
+	DefaultPublisher.Close()
+}
+
+// Event represents an Audit Log event.
+type Event struct {
```

Comment:
Can the existing `Event` struct not be used?

Reply:
Why do we would not want to use it?

Using it allows to have a similar behavior as the JS SDK, it also self document how an event can be filled (it would be even better if we can add godoc above each field, explaining what it is and give directions on how to fill the values).

Reply:
Sorry I might have misunderstood this one. The reason I did not want to reuse the current implementation is because:
- For now the action of this implementation does not use an Action type
- The current event implementation contains a logic to publish the event. People tend to take a look on godoc to learn how a package can be use, having 2 different way to do the same thing is confusing.

Reply:
When this gets merged it wouldn't make sense to have the `auditlog` directory that currently exists. We could still carry over the convenience methods that exist on the `Event` struct while still publishing new events using this new approach.

For example, with the existing `Event` struct, other structs in an application can implement the `auditlog.Auditable` interface which makes it easy for applications to set the group, actor, and target for an event. From experience trying to set the IDs and human strings manually each time an action is logged is hard to maintain good data quality.

Reply:
The `Auditable` interface implies that the user implements the methods to their types to satisfy it.

Implementing the interface adds the work of adapting user models in order to satisfy our interface. Since they still must have to know what to return when implementing it, I feel it ends to the same result but with additional steps.

Automating how an event would be filled could also be done in a function that the user would implement following his personal needs without having to add a vendor logic to his data model.

Having 2 ways to achieve a same goal can be confusing.
When I'm reading the godoc, if I see that there is method to fill the Event, I will think this is the preferred way to use the package. It gonna lead me to implement something more heavy in order to go to the goal I seek.

Reply:
> Implementing the interface adds the work of adapting user models in order to satisfy our interface. Since they still must have to know what to return when implementing it, I feel it ends to the same result but with additional steps.

The `Auditable` interface is optional, but not required. The integrator can set `ActorName`, `ActorID`, `TargetName` and `TargetID` manually if they like.

> Automating how an event would be filled could also be done in a function that the user would implement following his personal needs without having to add a vendor logic to his data model.

Totally! Integrators can write their own functions should they choose.

> Having 2 ways to achieve a same goal can be confusing.

The interface is used for the convenience functions `Event.SetActor`, `Event.SetTarget`, and if it's ever a value in the metadata. From experience it's difficult to make sure the same value is logged for the same model across many engineering teams creating and maintaining Audit Log events.

For example, without an interface, if my app had events like `team.create`, `team.update`, `team.destroy`, `team.add_member`, `team.remove_member`, `repository.update_team_permission`, `team_discussion.create`, etc. and I wanted to attach the team name and team ID for each Audit Log event it'd have to be:

```go
event.Action = "team.create"
event.Metadata["team"] = team.NameWithOwner # => org-name/team-name
event.Metadata["team_id"] = team.ID # => 1
```

Repeating this for every event that wants to attach the team is repetitive and error prone. If another team adds a `team_discussion.destroy` event later down the road and they populate the team manually with the wrong function or field like so:

```go
event.Action = "team.create"
event.Metadata["team"] = team.Name # => team-name
event.Metadata["team_id"] = team.ID # => 1
```

It may not be caught in code review and the team name logged across all Audit Log actions will be inconsistent and could confuse administrators. The interface gives teams the opportunity to encapsulate the logic necessary in one place and make it testable.

```go
func (t Team) ToAuditableName() string {
  return fmt.Sprintf("%s/%s", t.OrganizationName, t.Name)
}

func (t Team) ToAuditableID() string {
  return strconv.Itoa(t.DatabaseID)
}

event.Action = "team.create"
event.Metadata["team"] = team
```

Which will result in:

```json
{
  "team": "org-name/team-name",
  "team_id": "1"
}
```

> I think the first step to go toward data quality is providing good documentation. Adding godoc above each fields explaining what it is and how to fill it could greatly contribute to data quality.

💯 we should definitely add documentation for each field when possible.

Reply:
I agree with @maxence-charriere in regards to simplifying the sdk.

These setters add additional implementation requirements for consumers. They  will have to redesign all of their current data models in-order to use this interface.

This decision seems to only impact larger organizations where multiple teams have difficulty synchronizing a Target, Action, or Actor name.

I think our out-of-the box solution should not force a small organization, with no synchronization problems to sit in the same boat as a large org.

Reply:
> These setters add additional implementation requirements for consumers. They will have to redesign all of their current data models in-order to use this interface.

The setters are optional, not required. Same with the `Auditable` interface. The `Auditable` interface isn't exclusive for only the actor and target fields, it is also applicable for any struct that gets included in the metadata. That could be up to 50 properties.

I haven't suggested that teams are required to use the `SetActor` or `SetTarget` functions, I'm mostly trying to highlight what the SDK and eventual data would look like without an interface that small to large companies _should_ implement.

The interface lets the SDK know how each struct should be represented for the human readable string and for the machine readable identifier. Teams implementing the SDK could still construct an `Event` struct manually and set each field as demonstrated in this PR. Once someone sets up 10 unique events in the Audit Log they will inevitably start sharing structs across events, and you risk data quality issues such as mixed key names representing the same value across events (for example `org` in `team.remove_member` and `organization` in `team.add_member`), inconsistent values logged for the same property (for example for a team it would be `workos-inc/employees` in one event and just `employees` in another).

For example, on a `team.add_member` event with no metadata we either have to set the target as the user being added or the team the user is being added to:

```json
{
  "actor": "dewski",
  "actor_id": "User;79995",
  "target": "maxchehab",
  "target_id": "User;12345",
  "metadata": {}
}
```

or:

```json
{
  "actor": "dewski",
  "actor_id": "User;79995",
  "target": "workos-inc/employees",
  "target_id": "Team;12345",
  "metadata": {}
}
```

To start with we'll at least have to add the `team` or `user` to metadata to get a clearer picture of what happened, so I'll just stick with the target being the user being added as the base event:

```json
{
  "actor": "dewski",
  "actor_id": "User;79995",
  "target": "maxchehab",
  "target_id": "User;12345",
  "metadata": {
    "team": "workos-inc/employees",
    "team_id": "Team;12345"
  }
}
```

If we need to know the organization that the team belonged to at the time of the event we'll need to add that:

```json
{
  "actor": "dewski",
  "actor_id": "User;79995",
  "target": "maxchehab",
  "target_id": "User;12345",
  "metadata": {
    "team": "workos-inc/employees",
    "team_id": "Team;12345",
    "organization": "workos-inc",
    "organization_id": "Organization;12345",
    "business": "workos-inc",
    "business_id": "Business;12345"
  }
}
```

As is the Go struct would look something like this:

```go
event := auditlog.Event{
  Action: "team.add_member",
  Actor: actor.Login,
  ActorID: actor.ID,
  Target: user.Login,
  TargetID: user.ID,
  Metadata: map[string]interface{}{
    "team": fmt.Sprintf("%s/%s", organization.Login, team.Slug),
    "team_id": team.ID,
    "organization": organization.Login,
    "organization_id": organization.ID,
    "business": business.Login,
    "business_id": business.ID,
  }
}
```

If someone wanted to add a `team.remove_member` event a few months later they'll either have to know to copy and paste this struct as a starting point, or build it out by hand and may not know the exact values to be used. If every time someone is expected to add an event, values will start to drift since there isn't anything telling them what they should be.

```go
event := auditlog.Event{
  Action: "team.remove_member",
  Actor: actor.Login,
  ActorID: actor.ID,
  Target: user.Login,
  TargetID: user.ID,
  Metadata: map[string]interface{}{
    "team": team.Name,
    "teamId": team.ID,
    "org": organization.Name,
    "orgId": organization.ID,
    "business": business.Name,
    "businessId": business.ID,
  }
}
```

_Take note of the inconsistent key names, `team`/`org`/`business` values._

If I'm to ever consume this data pragmatically I'll have to always account for the organization being stored in an `org` and `orgId` field. While you can't _prevent_ it from happening on purpose, I believe the SDK should do its best to try and prevent that from happening.

My only suggestion has been to encourage the use of interfaces to prevent drifting data, consistent key names, and values. The alternative way to write these events would be:

```go
event := auditlog.Event{
  Action: "team.remove_member",
  Actor: actor.Login,
  ActorID: actor.ID,
  Target: user.Login,
  TargetID: user.ID,
  Metadata: map[string]interface{}{
    "team": team,
    "organization": organization,
    "business": business,
  }
}
```

It makes it less ambiguous what the values should be and reduce the chance of data quality issues. While large teams could create their own interface and functions to generate events, it's putting a lot of work on each company integrating with WorkOS to roll it by hand.

I think the SDK should provide helper utilities such as the interface to get the most out of the tool. I didn't touch into other fields that you'd want to log across each event like `request_id`, `http_method`, `http_path`, etc. If I had to set those manually each time I added a new event it'd require more work on my part and more room for data quality issues where they could be missing completely and make it difficult to debug issues in the future.

> I think our out-of-the box solution should not force a small organization, with no synchronization problems, to sit in the same boat as a large org.

Small teams will eventually run into this as they start to grow and get more out of the WorkOS Audit Log.